AlphaWallet/Common/Services/ContentLoader.swift

```diff
@@ -0,0 +1,233 @@
+    private var publishers: [URLRequest: LoadContentPublisher]
+    private let session: URLSession
+    private let cache: ContentCacheStorage
+    private let decoder: ImageDecoder
+    private let interceptor: ContentLoaderResponseInterceptor
+
+    init(publishers: [URLRequest: LoadContentPublisher] = [:],
+         session: URLSession,
+         cache: ContentCacheStorage,
+         decoder: ImageDecoder = ImageDecoder(),
+         interceptor: ContentLoaderResponseInterceptor) {
+
+        self.decoder = decoder
+        self.interceptor = interceptor
+        self.cache = cache
+        self.publishers = publishers
+        self.session = session
+    }
+
+    public func fetch(_ url: URL) -> LoadContentPublisher {
+        let request = URLRequest(url: url)
+        return fetch(request)
+    }
+
+    public func fetch(_ urlRequest: URLRequest) -> LoadContentPublisher {
+        return buildFetchPublisher(urlRequest)
+            .prepend(.loading)
+            .eraseToAnyPublisher()
+    }
+
+    private func buildFetchPublisher(_ urlRequest: URLRequest) -> LoadContentPublisher {
+        guard let url = urlRequest.url else { return .fail(.invalidUrl) }
+
+        if let data = try? cache.value(for: ImageCacheKey.cacheKey(for: url, prefix: .raw)), let data = try? decoder.decode(data: data) {
+            switch data {
+            case .image, .svg:
+                return .just(.done(data))
+            case .video(let video):
+                let preview = try? cache.value(for: ImageCacheKey.cacheKey(for: url, prefix: .videoPreview)).flatMap { UIImage(data: $0) }
+                let video = Video(url: video.url, preview: preview)
+
+                return .just(.done(.video(video)))
+            }
+        }
+
+        return Just(urlRequest)
+            .receive(on: queue)
+            .setFailureType(to: ContentLoaderError.self)
+            .flatMap { [weak self, queue, session, cache, decoder, interceptor] urlRequest -> LoadContentPublisher in
+                if let publisher = self?.publishers[urlRequest] {
+                    return publisher
+                } else {
+                    let publisher = session.dataTaskPublisher(for: urlRequest)
+                        .receive(on: queue)
+                        .mapError { ContentLoaderError.internal($0) }
+                        .flatMap { response -> LoadContentPublisher in
+                            do {
+                                guard let resp = response.response as? HTTPURLResponse else {
+                                    throw ContentLoaderError.invalidData
+                                }
+                                let value = try decoder.decode(response: resp, data: response.data)
+                                try cache.set(data: value.data, for: ImageCacheKey.cacheKey(for: url, prefix: .raw))
+
+                                return interceptor.intercept(response: value.image)
```

Comment:
This line doesn't read very well because we can't tell what it's doing. Do you forsee we need the extensibility? Otherwise this would be better right?

```
return generateStillIfVideo(response: value: image)
```

Still not so good, but at least it's not very generic like "intercept" which doesn't say what it's doing.